demos/demo-content/main.go

```diff
@@ -0,0 +1,198 @@
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/GoogleContainerTools/kpt/internal/printer"
+	"github.com/GoogleContainerTools/kpt/pkg/content/open"
+	"github.com/GoogleContainerTools/kpt/pkg/location"
+)
+
+func main() {
+	// wire the global printer
+	pr := printer.New(os.Stdout, os.Stderr)
+
+	// create context with associated printer
+	ctx := printer.WithContext(context.Background(), pr)
+
+	// run through samples on location.Dir{}
+	fmt.Print("\n## Temp folder example\n\n")
+	if err := tempFolderExample(ctx); err != nil {
+		fmt.Printf("error in tempFolderExamples: %v\n", err)
+	}
+
+	// run through examples on location.Git{}
+	fmt.Print("\n## GitHub blueprint example\n\n")
+	if err := githubBlueprintExample(ctx); err != nil {
+		fmt.Printf("error in githubBlueprintExample: %v\n", err)
+	}
+
+	if len(os.Args) == 2 {
+		// run through examples on provided location
+		fmt.Print("\n## Parsed location example\n\n")
+		if err := locationExample(ctx, os.Args[1]); err != nil {
+			fmt.Printf("error in locationExample: %v\n", err)
+		}
+	}
+}
+
+var helloYaml = `
+apiVersion: example.com/v1
+kind: Greeting
+metadata:
+  name: hello
+spec:
+  target: world
+`
+
+func tempFolderExample(ctx context.Context) error {
+	// set up temp dir
+	tmp, err := os.MkdirTemp("", "")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmp)
+
+	if err := os.WriteFile(filepath.Join(tmp, "README.md"), []byte("Hello world"), os.ModePerm); err != nil {
+		return err
+	}
+	if err := os.WriteFile(filepath.Join(tmp, "hello.yaml"), []byte(helloYaml), os.ModePerm); err != nil {
+		return err
+	}
+
+	return runAll(
+		ctx,
+		location.Dir{
+			Directory: tmp,
+		},
+	)
+}
+
+func githubBlueprintExample(ctx context.Context) error {
+	return runAll(
+		ctx,
+		location.Git{
+			Repo:      "https://github.com/GoogleCloudPlatform/blueprints",
+			Directory: "catalog/gke",
+			Ref:       "main",
+		},
+	)
+}
+
+func locationExample(ctx context.Context, arg string) error {
+	ref, err := location.ParseReference(
+		arg,
+		location.WithContext(ctx),
+		location.WithParsers(location.OciParser, location.GitParser, location.DirParser),
+	)
+	if err != nil {
+		return err
+	}
+
+	return runAll(ctx, ref)
+}
+
+func runAll(ctx context.Context, ref location.Reference) error {
+	if err := openContent(ctx, ref); err != nil {
+		return fmt.Errorf("open content: %v", err)
+	}
+	if err := readFromFileSystem(ctx, ref); err != nil {
+		return fmt.Errorf("open content as FileSystem: %v", err)
+	}
+	if err := readFromReader(ctx, ref); err != nil {
+		return fmt.Errorf("open content as Reader: %v", err)
+	}
+	if err := readFromFS(ctx, ref); err != nil {
+		return fmt.Errorf("open content as FS: %v", err)
+	}
+	return nil
+}
+
+func openContent(ctx context.Context, ref location.Reference) error {
+	// open it as content
+	src, err := open.Content(ref, open.WithContext(ctx))
```

Comment:
having `open` as the package name is little awkward. I am almost thinking if we can do this `content.Open(...)`

Reply:
That convention does have an odd first impression, doesn't it? With only one "verbish" go sub package to start it's hard to tell if it's paying for itself.

The place that pattern comes from is in the oci lib --- for example instead of `images.Validate(img, opts ...images.ValidateOption)` there's a `validate.Image(img, opts ...validate.Option)`

One way it really shines is that the option factories for a verbish group are very focused to what apply to methods in that group. e.g. `validate.Image(img, validate.Fast)` where Fast is of type `validate.Option`... Other examples remote.Image(name, ...remote.Option)`, daemon.Image, daemon.Write, tarball.Image, etc...

---

but looping back to Kpt apis specifically, where you see that is `src := open.Content(reference)` to open as content,`src := open.FileSystem(reference)` to open but with filesystem as access choice, `src := open.Reader(reference)` to open but with reader as access choice...

in the base package there is `fs := content.FileSystem(src)` and `r := content.Reader(src)` also - but everything you'll expect to find there is "I've got a content as the argument, and the methods often relate to a return value or property. Frequently something which would be optional and duck typed but default behavior is provided"

Say! That's the other example. `io/fs` kind of acts this way where `x := fs.ReadFile(fsys fs.FS, path string)` where the package name "fsXxx." kind of means "these are methods you may find on a FS, passed as the first argument" like "content.Xxx" is a method that you may find on a Content, which you pass as the first argument.

But the way you get your hands on a "FS" is somewhere else, like `os.DirFS(path)`